infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/log/LogWatcher.java

```diff
@@ -0,0 +1,162 @@
+  private final String workspaceId;
+  private final Set<String> podsOfInterest;
+
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+
+  /**
+   * Map of current watchers where key is name of the container and value is {@link
+   * ContainerLogWatch} instance.
+   */
+  private final Map<String, ContainerLogWatch> currentContainerWatchers = new ConcurrentHashMap<>();
+
+  public LogWatcher(
+      KubernetesClientFactory clientFactory,
+      String workspaceId,
+      String namespace,
+      Set<String> podsOfInterest,
+      Executor executor,
+      LogWatchTimeouts timeouts)
+      throws InfrastructureException {
+    this.client = clientFactory.create(workspaceId);
+    this.workspaceId = workspaceId;
+    this.namespace = namespace;
+    this.containerWatchersThreadPool = executor;
+    this.timeouts = timeouts;
+    this.podsOfInterest = podsOfInterest;
+  }
+
+  public void addLogHandler(PodLogHandler handler) {
+    logHandlers.add(handler);
+  }
+
+  @Override
+  public void handle(PodEvent event) {
+    final String podName = event.getPodName();
+
+    final String containerName = event.getContainerName();
+    if (containerName != null && event.getReason().equals(STARTED_EVENT_REASON)) {
+      for (PodLogHandler logHandler : logHandlers) {
+        // we need to synchronize here so we won't add new watcher while we're cleaning them
+        synchronized (this) {
+          if (closed.get()) {
+            return;
+          }
+          if (podsOfInterest.contains(podName)
+              && !currentContainerWatchers.containsKey(podContainerKey(podName, containerName))) {
+            ContainerLogWatch logWatch =
+                new ContainerLogWatch(
+                    client, namespace, podName, containerName, logHandler, timeouts);
+            currentContainerWatchers.put(podContainerKey(podName, containerName), logWatch);
+            LOG.trace(
+                "adding [{}] to watching containers now watching [{}]",
+                containerName,
+                currentContainerWatchers.keySet());
+            containerWatchersThreadPool.execute(logWatch);
+          } else {
+            LOG.debug(
+                "Not for this handler or already watching '{} : {} : {}'",
+                namespace,
+                podName,
+                containerName);
+          }
+        }
+      }
+    }
+  }
+
+  private String podContainerKey(String podName, String containerName) {
+    return podName + ":" + containerName;
+  }
+
+  public void close() {
+    this.close(false);
+  }
+
+  /**
+   * Closes all opened log watchers. In case of failed workspace, we want to block the pod for some
+   * time before removing it so we have better chance to get all the logs from it. If that's the
+   * case, use {@code needWait=false}. Otherwise watchers will be cleaned immediately, which does
+   * not ensure that we get all the logs.
+   *
+   * @param needWait true if we need to block before cleanup
```

Comment:
I think `pause`  or `delay` might be clearer than `block` here.

Reply:
fixed